R/create_single_dose_dataset.R

```diff
@@ -0,0 +1,272 @@
+#'
+#' @param start_date The start date
+#'
+#'   A date or date-time object is expected.
+#'
+#'   Default: `ASTDT`
+#'
+#' @param end_date The end date
+#'
+#'   A date or date-time object is expected.
+#'
+#'   Default: `AENDT`
+#'
+#' @param lookup_table The dose frequency value lookup table
+#'
+#'   The table used to look up `dose_freq` values and determine the appropriate
+#'   multiplier to be used for row generation. If a lookup table other than the
+#'   default is used, it must have columns `DoseWindow`, `DoseCount`, and
+#'   `DayConversionFactor`
+#'
+#'   Default: `dose_freq_lookup`
+#'
+#'  Permitted Values for `DoseWindow`: "DAY", "WEEK", "MONTH", "YEAR"
+#'
+#' @param lookup_column The dose frequency value column in the lookup table
+#'
+#'   The column of `lookup_table`.
+#'
+#'   Default: `CDISCSubmissionValue` (column of `dose_freq_lookup`)
+#'
+#'
+#' @details Each aggregate dose row is split into multiple rows which each represent
+#' a single dose.The number of completed dose periods between `start_date` and
+#' `end_date` is calculated with `compute_duration` and multiplied by the count
+#' of doses per period. `start_date` and `end_date` are then adjusted by an
+#' appropriate increment using `DayConversionFactor`.
+#'
+#' @author Michael Thorpe, Andrew Smith
+#'
+#' @return The input dataset with a single dose per row.
+#'
+#' @export
+#'
+#' @examples
+#' data <- tibble::tribble(
+#'   ~USUBJID, ~EXDOSFRQ, ~ASTDT, ~AENDT,
+#'   "P01", "Q2D", lubridate::ymd("2021-01-01"), lubridate::ymd("2021-01-07"),
+#'   "P01", "Q3D", lubridate::ymd("2021-01-08"), lubridate::ymd("2021-01-15"),
+#'   "P01", "EVERY 2 WEEKS", lubridate::ymd("2021-01-15"), lubridate::ymd("2021-01-29")
+#' )
+#'
+#' create_single_dose_dataset(data, by_vars = vars(USUBJID))
+create_single_dose_dataset <- function(dataset,
+                                       by_vars,
+                                       dose_freq = EXDOSFRQ,
+                                       start_date = ASTDT,
+                                       end_date = AENDT,
+                                       lookup_table = dose_freq_lookup,
+                                       lookup_column = CDISCSubmissionValue) {
+
+  col_names <- colnames(dataset)
+  by_vars <- assert_vars(by_vars)
+  dose_freq <- assert_symbol(enquo(dose_freq))
+  lookup_column <- assert_symbol(enquo(lookup_column))
+  start_date <- assert_symbol(enquo(start_date))
+  end_date <- assert_symbol(enquo(end_date))
+  assert_data_frame(dataset, required_vars = quo_c(by_vars, dose_freq, start_date, end_date))
+  assert_data_frame(lookup_table, required_vars = vars(DoseWindow, DoseCount, DayConversionFactor))
+
+    # Set up lookup table to be joined to dataset
+
+    lookup <- lookup_table %>%
+      rename(!!dose_freq := !!lookup_column)
+
+    # Use compute_duration to determine the number of completed dose periods
+
+    dataset <- dataset %>%
+      left_join(lookup, by = as_name(dose_freq)) %>%
```

Comment:
Think I'm in a bit of a bind unless folks have sharp ideas: `as_name` is not in the `rlang` version for R 3.5 it looks like? But if I change it to `quo_name`, I believe that gets me a warning, and going straight to `as.character()[2]` also makes R mad.